### ISSUE TITLE
documentation

### ISSUE THREAD

**Issue by MonsieurCactus** [ghcjs/ghcjs#568]
I have found the documentation a little bit sparse.  After asking a question or two on stackOverflow it has become clear I do not understand the purpose of ghcjs.

> GHCJS is a Haskell to JavaScript compiler that uses the GHC API.

I took this to mean this was a version of Haskell that applies to front-end web development, which is wrong.      Some people suggested ghcjs-dom or reflex-frp.  If there was more documentation, I might have less trouble.

Then someone on StackOverflow suggested I contribute my own documentation, since I m not happy with the current state of things.   How can I get started understanding better how ghcjs is used in general and in the particular cases I am interested in?

**Comment by 3noch** [comment]
@MonsieurCactus For the sake of documentation, GHCJS is identical to GHC except that it produces JavaScript code instead of machine code. With GHCJS, most things that you might write in Haskell to run on a server can also run inside a browser. But this also opens the door for more JavaScript-oriented libraries like ghcjs-dom and reflex-dom. These specifically target the browser DOM and run inside the browser. This is not possible with GHC since it produces machine code.

**Comment by cdsmith** [comment]
> I took this to mean this was a version of Haskell that applies to front-end web development, which is wrong. 

I don't think this is wrong, so I'm curious where you got this idea.  GHCJS compiles Haskell to JavaScript, and that JavaScript absolutely can (and usually does) run in web front-ends.  The other projects (`ghcjs-dom` and `reflex-frp`) are *libraries* which are commonly used *with* GHCJS to accomplish that goal.

**Comment by MonsieurCactus** [comment]
@3noch what is the correct ghcjs library can be used to target the browser DOM?

@cdsmith Here is the StackOverflow question [[1](http://stackoverflow.com/questions/42254610/how-to-code-h1-tags-with-ghcjs/42262597)], please excuse me I am getting a bit confused and irate.  Basically I am trying to build a "hello world" type example to use ghcjs to target the DOM (i.e. build web sites) and we touch upon the scarcity of examples and documentation.  And maybe there is stuff I am just not getting about ghcjs (this is my first time).  

*Obviously* I want to build something more sophiticased, but my example just uses an `h1` and a `p` tag.  

In contract, a language like [Elm](http://elm-lang.org/) makes its connection to web development very clear and very explicit and they go through make pains to make sure it is easy to follow.  It does have limitation since Elm's types are not as rich.  However, I may have to stick to Elm since I don't understand how to take the first steps towards ghcjs.

**Comment by 3noch** [comment]
@MonsieurCactus There are several libraries to target the DOM. `ghcjs-dom` is a low-level library that works but it's rough to work with (just like using raw DOM functions in JS). Reflex-DOM is an FRP system built on this which is much easier to use, but still rather advanced. React-Haskell (https://hackage.haskell.org/package/react-haskell) provides bindings to React. There are other libraries as well.

If you're coming from something like Elm then you need to know a few things about GHCJS:
  * GHCJS is full-blown Haskell - so the language is not specifically designed with web development in mind. This isn't a bad thing but it's different.
  * GHCJS is young - Elm is young too but it has a much less ambitious goal. It *only* targets front-end web pages. GHCJS targets the entire Haskell ecosystem and can even run server side apps in JS via node.
  * GHCJS is still in early development - Elm is in a nice, clean, well documented box shipped to your front door. GHCJS is much less polished than that. Clearly it has all the advantages of having Haskell's many powerful language features and almost all of the Haskell ecosystem. Still, it's harder to get started. [Reflex-Platform](https://github.com/reflex-frp/reflex-platform) is probably one of the easiest ways to get started. Stack also supports building with GHCJS but the support is also still young.

If you want to whip up something quick then GHCJS is likely not for you. If you plan to have a lot of Haskell code in your code base and you want to invest in the future, then GHCJS is the right choice.

**Comment by MonsieurCactus** [comment]
@3noch Thank you for agreeing with me that GHCJS is young and under development.  I've been getting contradictory feedback even on very basic points.  That's why I am asking here.  From StackOverflow:

> ghcjs isn't new at all, and people have already figured it out. Are you perhaps trying to say "I'm very new to ghcjs, and am still trying to figure out how it works"? 

I think your responses hit it on the head.  Elm is great for whipping up quick, beautiful functional web-sites.  If I need full-blown functional programming (e.g. running a server-side app in the browser), I may need to start learning Haskell.  

The nice part about these new languages, is that my experiences may have impact for the rest of the community.  This is why I started an issue on documentation since ghcjs (or ghcjs-dom) lacks very good [on-boarding](https://en.wikipedia.org/wiki/Onboarding), but this is an opportunity to contribute.  

The app I initially write will have an enormous amount of overhead, to write apps that are quite trivial.  Hopefully with the goal of something more substantial later on.  And the community may benefit from me documenting that experience.

**Comment by 3noch** [comment]
@MonsieurCactus Glad to be helpful. Yes, the term "young" is a bit overloaded. I don't mean to say that GHCJS is "young" in a time-line sense, but it is most certainly young in a "development" sense. Yes, it is being used by many companies in production. But that doesn't mean it's anywhere near the level of maturity in terms of documentation and community that React and AngularJS have.

If you don't already know Haskell then I suggest you start with building some simple Haskell apps with normal GHC and [stack](https://docs.haskellstack.org/en/stable/README/). GHCJS will likely provide enough onboarding frustration that you will want to take the journey in smaller steps. Regardless how you end up doing it, we will all be very thankful for any contributions you can offer along the way to make it easier for others like yourself.

**Comment by cdsmith** [comment]
@MonsieurCactus I think we just misunderstood each other.  GHCJS is definitely a tool for front-end web development with Haskell.  But it's not the *only* tool you need.  It's one of the lower-level parts, and you still have choices about how to use it.

The most straight-forward way to add an `h1` element to a document using GHCJS, for example, would be to write this:

    foreign import javascript unsafe "document.createElement($1)"
      js_createElement :: JSString -> IO JSVal
    foreign import javascript unsafe "$1.innerText = $2;"
      js_setInnerText :: JSVal -> JSString -> IO ()
    foreign import javascript unsafe "document.body.append($1);"
      js_appendToBody :: JSVal -> IO ()
    main = do
      h1 <- js_createElement "h1"
      js_setInnerText h1 "Some Header Here"
      js_appendToBody h1

But in practice, this isn't how anyone would do it!  Instead, they would use some library, which does this stuff for them.  GHCJS itself is just a compiler, and doesn't have an opinion about which library you would use to do this.  But if you're a fan of the FRP (functional reactive programming) approach, then `reflex-dom` is a good choice.  If you want a basic imperative model instead, then `ghcjs-dom` will do it for you.  (Typically, if you're programming with `ghcjs-dom`, you would not build the static part of your page with GHCJS at all; you'd write an HTML file yourself, and use GHCJS to implement dynamic logic that changes bit of that page, instead.)

Also agreed that there's room for plenty of improvement in documentation, and (more importantly) less painful steps to get started with GHCJS.  So I wan't trying to argue against the bug.  Just trying to understand where you got the idea that GHCJS isn't for web front-ends.  It is; it's just a low-level tool.  Like a hammer, it does a key job, but doesn't really have any opinions on how you should design the house.

**Comment by MonsieurCactus** [comment]
@cdsmith For "Hello World" those low-level functions may be a correct place to start, since I have not written anything yet! `ghcjs-dom` has even *less* documentation (read none) on GitHub.  

GHCJS lacks a good **onboarding** process, the crucial information that me or any newcomer needs to get started.  Including important choices of architecture, that a programmer may not even be aware of.

* `reflex-dom` for Functional Reactive Programming
* `ghcjs-dom` if I am not committed to FRP

It sounds like my next step may be to raise an issue other there.

**Comment by harryxp** [comment]
@MonsieurCactus Thank you for raising the question.  I had the same issue figuring out how to start, but eventually got it to work.  I have documented my process [here on github](https://github.com/harryxp/ghcjs-playground).  Let me know if it helps!
